streaming/base/converters/dataframeToMDS.py

```diff
@@ -0,0 +1,250 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A utility to convert databricks' tables to MDS."""
+
+import json
+import os
+import shutil
+from argparse import ArgumentParser
+from collections.abc import Iterable
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, Union
+
+import pandas as pd
+from pyspark import TaskContext
+from pyspark.sql import SparkSession
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.types import StringType, StructField, StructType
+
+from streaming import MDSWriter
+from streaming.base.storage.upload import CloudUploader
+
+default_mds_kwargs = {
+    'compression': 'zstd:7',
+    'hashes': ['sha1', 'xxh64'],
+    'size_limit': 1 << 27,
+    'progress_bar': 1,
+    'columns': {
+        'tokens': 'bytes'
+    },
+    'keep_local': False
+}
+
+default_udf_kwargs = {
+    'concat_tokens': 2048,
+    'tokenizer': 'EleutherAI/gpt-neox-20b',
+    'eos_text': '<|endoftext|>',
+    'compression': 'zstd',
+    'split': 'train',
+    'no_wrap': False,
+    'bos_text': '',
+    'key': 'content',
+}
+
+
+def is_iterable(obj: Any) -> bool:
+    """Check if obj is iterable."""
+    return issubclass(type(obj), Iterable)
+
+
+def do_merge_index(partitions: Iterable,
+                   mds_path: Union[str, Tuple[str, str]],
+                   keep_local: bool = True,
+                   skip: bool = False):
+    """Merge index.json from partitions into one for streaming."""
+    print('I am here 1', skip)
+    if not partitions or skip:
+        return
+
+    shards = []
+
+    for row in partitions:
+        mds_partition_index = f'{row.mds_path}/index.json'
+        mds_partition_basename = os.path.basename(row.mds_path)
+        obj = json.load(open(mds_partition_index))
+        for i in range(len(obj['shards'])):
+            shard = obj['shards'][i]
+            for key in ['raw_data', 'zip_data']:
+                if shard.get(key):
+                    basename = shard[key]['basename']
+                    obj['shards'][i][key]['basename'] = os.path.join(mds_partition_basename,
+                                                                     basename)
+        shards += obj['shards']
+
+    obj = {
+        'version': 2,
+        'shards': shards,
+    }
+
+    if isinstance(mds_path, str):
+        mds_index = os.path.join(mds_path, 'index.json')
+    else:
+        mds_index = os.path.join(mds_path[0], 'index.json')
+
+    with open(mds_index, 'w') as out:
+        json.dump(obj, out)
+
+
+def dataframeToMDS(dataframe: DataFrame,
+                   out: Union[str, Tuple[str, str]],
+                   columns: Dict[str, str],
```

Comment:
Why can't the type of the columns be inferred by the dataframe if not specified. Having the user specify all the column types manually risks columns being lost in the conversion

Reply:
In practice, inferring schema from table is very error prone. It's users' responsibility to specify correct input.

Reply:
One example is that a user id column can be interpreted as both integer or str.

Reply:
user should provide correct dataframe and specify correct schema that they want. It avoids a lot of runtime issues. MDSwriter also expects columns to be specified.

Reply:
yes, it's possible. I enabled that. From what I seen before, inferring schema from dataframe is error prone when table has more complicated structure.